lib/client/conntest/ssh.go

```diff
@@ -49,30 +63,243 @@ func NewSSHConnectionTester(cd services.ConnectionsDiagnostic) *SSHConnectionTes
+	}
+
+	connectionDiagnosticID := uuid.NewString()
+	connectionDiagnostic, err := types.NewConnectionDiagnosticV1(connectionDiagnosticID, map[string]string{},
 		types.ConnectionDiagnosticSpecV1{
-			Message: types.DiagnosticMessageWaiting,
+			// We start with a failed state so that we don't need an extra update when returning non-happy paths.
+			// For the happy path, we do update the Message to Success.
+			Message: types.DiagnosticMessageFailed,
 		})
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	if err := s.cd.CreateConnectionDiagnostic(ctx, connectionDiagnostic); err != nil {
+	if err := s.userClt.CreateConnectionDiagnostic(ctx, connectionDiagnostic); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	key, err := client.GenerateRSAKey()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	currentUser, err := s.userClt.GetCurrentUser(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	certs, err := s.userClt.GenerateUserCerts(ctx, proto.UserCertsRequest{
+		PublicKey:              key.MarshalSSHPublicKey(),
+		Username:               currentUser.GetName(),
+		Expires:                time.Now().Add(time.Minute).UTC(),
+		ConnectionDiagnosticID: connectionDiagnosticID,
+	})
+	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	// TODO(marco): test ssh connection
-	// - create certificate for user with extra field
-	// - create ssh client using that certificate
-	// - if connection fails because of a network error, a trace must be included indicating that the host is not reachable
-	// - other traces will be added by the Node itself (rbac checks, principal)
-	connectionDiagnostic.SetMessage("dry-run")
-	connectionDiagnostic.SetSuccess(true)
+	key.Cert = certs.SSH
+	key.TLSCert = certs.TLS
+
+	certAuths, err := s.userClt.GetCertAuthorities(ctx, types.HostCA, false /* loadKeys */)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	hostkeyCallback, err := hostKeyCallbackFromCAs(certAuths)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	key.TrustedCA = auth.AuthoritiesToTrustedCerts(certAuths)
+
+	keyAuthMethod, err := key.AsAuthMethod()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	clusterName, err := s.userClt.GetClusterName()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	clientConfTLS, err := key.TeleportClientTLSConfig(nil, []string{clusterName.GetClusterName()})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	host := req.ResourceName
+	hostPort := 0
+	tlsRoutingEnabled := req.TLSRoutingEnabled
+
+	parsedProxyHostAddr, err := client.ParseProxyHost(req.ProxyHostPort)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	webProxyAddr := parsedProxyHostAddr.WebProxyAddr
+	sshProxyAddr := parsedProxyHostAddr.SSHProxyAddr
+
+	key.KeyIndex = client.KeyIndex{
+		Username:    req.SSHPrincipal,
+		ProxyHost:   webProxyAddr,
+		ClusterName: clusterName.GetClusterName(),
+	}
+
+	if !tlsRoutingEnabled {
```

Comment:
Why does this need to be done only when TLS routing is not enabled?

Also, I'd factor this out into a separate function e.g. `getNodePort` or something.

Reply:
Extracted to a method :+1: 

We actually only need the node's hostname and pass it to the `TeleportClient` as the `Host` param.

I'll probably set up a Teleport Cluster where we don't use TLSRouting in order to manually test this (unit tests are passing though).